### ISSUE TITLE
Not possible to use multidimensional arrays in widget parameters

### ISSUE THREAD

**Issue by ilnytskyi** [magento/magento2#19909]
### Summary (*)

Magento 
2.2.*, 2.2.5, 2.2.6, 2.2.7

Not possible to use multidimensional arrays in widget parameters
When add custom field set as a **block** parameter type to **widget.xml**

### Examples (*)

Add custom parameter type (in my case it is an options panel)
![selection_007](https://user-images.githubusercontent.com/3192770/50291012-a3f1f400-046d-11e9-9f35-4fe2a9bc1f67.png)

Use parameters names and keys like this 
```
parameters[groups_options][order][option[1]: 0
parameters[groups_options][title][option][1][store][0]: store_label_admin
parameters[groups_options][title][option][1][store][1]: store_label_1
parameters[groups_options][title][option][1][store][2]: store_label_2
```
See an error 
![selection_008](https://user-images.githubusercontent.com/3192770/50291148-0814b800-046e-11e9-814c-8a568db4e7e1.png)

in this file the code looks like this 
`magento/vendor/magento/module-widget/Model/Widget/Instance.php:609`

```
            if ($name == 'conditions') {
                $name = 'conditions_encoded';
                $value = $this->conditionsHelper->encode($value);
            } elseif (is_array($value)) {
                $value = implode(',', $value); 
            }
```
It is not really possible to reproduce on vanilla instance but we can try to substitute a request parameters for existing widget fields 

### Proposed solution

```
// can't we simply serialize an input here instead of implode it ?
// anyway in the database 'widget_instance' all widget values are included in json 
 $value = implode(',', $value); 
```
![selection_009](https://user-images.githubusercontent.com/3192770/50291358-af91ea80-046e-11e9-8da7-3b07246b341f.png)

**Comment by magento-engcom-team** [comment]
Hi @ilnytskyi. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@ilnytskyi do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @maheshWebkul721. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
Hi @eduard13. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@eduard13 Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-97290` were created

**Comment by eduard13** [comment]
Hi @ilnytskyi, these days I have also encountered the same issue while adding a dynamic rows field to a custom widget. In your case, I suggest to use `conditions` as field name, at least for me it helped.

**Comment by ilnytskyi** [comment]
@eduard13 thanks for this suggestion. I tired conditions but in my case I want to group them and give them different titles or other attributes.

**Comment by magento-engcom-team** [comment]
Hi @maheshWebkul721. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by magento-engcom-team** [comment]
Hi @eduard13. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by magento-engcom-team** [comment]
Hi @ilnytskyi. Thank you for your report. 
The issue has been fixed in magento/magento2#21008 by @eduard13 in 2.3-develop branch
Related commit(s):
 - [fa22418f6f0fd629b440c171e883d6242c0f32bb](https://github.com/magento/magento2/commit/fa22418f6f0fd629b440c171e883d6242c0f32bb)
 - [f35392661962f824a474a260618a9bb8ec41549e](https://github.com/magento/magento2/commit/f35392661962f824a474a260618a9bb8ec41549e)
 - [7eb567e48314c8e2f290838b3eefd23656e58eb7](https://github.com/magento/magento2/commit/7eb567e48314c8e2f290838b3eefd23656e58eb7)
 - [718bb76ae948b3dc7ddd2ec447fcb0aa291df55c](https://github.com/magento/magento2/commit/718bb76ae948b3dc7ddd2ec447fcb0aa291df55c)

The fix will be available with the upcoming 2.3.2 release.

**Comment by magento-engcom-team** [comment]
Hi @ilnytskyi. Thank you for your report. 
The issue has been fixed in magento/magento2#22214 by @ilnytskyi in 2.2-develop branch
Related commit(s):
 - [3a0383808ced01ac1d6af8d7df8fad334866b3d5](https://github.com/magento/magento2/commit/3a0383808ced01ac1d6af8d7df8fad334866b3d5)
 - [4c84817c2a3c5b3b4985900e024437f03ab5704b](https://github.com/magento/magento2/commit/4c84817c2a3c5b3b4985900e024437f03ab5704b)
 - [640127d884e8bf47206cc919f2af34b766b1222d](https://github.com/magento/magento2/commit/640127d884e8bf47206cc919f2af34b766b1222d)
 - [3a4150d0910f75cd884325cdbd13f1e731da47cb](https://github.com/magento/magento2/commit/3a4150d0910f75cd884325cdbd13f1e731da47cb)
 - [c365d6cb5def47773e9a06eebd196039b7b76d52](https://github.com/magento/magento2/commit/c365d6cb5def47773e9a06eebd196039b7b76d52)

The fix will be available with the upcoming 2.2.9 release.

**Comment by jevgenijmokrousov** [comment]
> Hi @ilnytskyi, these days I have also encountered the same issue while adding a dynamic rows field to a custom widget. In your case, I suggest to use `conditions` as field name, at least for me it helped.

As an option change Magento\Widget\Model\Widget\Instance.php: 607 to

```
            if ($name == 'conditions') {
                $name = 'conditions_encoded';
                $value = $this->conditionsHelper->encode($value);
            } elseif(count($value) !== count($value, COUNT_RECURSIVE)) { // multidimensional array
                $value = $this->conditionsHelper->encode($value);
            } elseif (is_array($value)) {
                $value = implode(',', $value);
            }
```
